File: src/Python/i_rda.py
# Copyright (c) 2022 Foldda Pty Ltd
# Licensed under the GPL License V3 -
# https://github.com/foldda/charian/blob/main/LICENSE

import abc

from typing import TYPE_CHECKING

# avoid circular import
if TYPE_CHECKING: 
    from rda import Rda


class IRda: 
    @abc.abstractmethod
    def to_rda(self)->'Rda':
        pass

    @abc.abstractmethod
    def from_rda(self, rda: 'Rda'):
        pass

File: src/Java/src/main/java/UniversalDataTransport/IRda.java
// Copyright (c) 2020 Michael Chen
// Licensed under the MIT License -
// https://github.com/foldda/rda/blob/main/LICENSE

package UniversalDataTransport;

/*
 * RDA-Serializable object implements this interface to ensure
 * 1) its properties can be stored into an resulting RDA object,
 * 2) used a provided RDA to restore this object's properties' values.
 *
 */

public interface IRda {
  /// <summary>
  /// Stores properties into the RDA.
  /// </summary>
  /// <returns>An Rda instance that carries the properties of this object.</returns>
  Rda ToRda();

  /// <summary>
  /// Populate properties with the values from the RDA
  /// </summary>
  /// <param name="rda">An Rda instance that carries the properties of an object to be restored.</param>
  IRda FromRda(Rda rda);
}
